File: array-pair-sum/array-pair-sum.js
// This is just a modification of the integer difference problem presented elsewhere
var arrayPairSum = function (k, array) {
  var hash  = {},
      pairs = [];

  // Loop through the array once, storing the results in an object for a
  // time complexity of O(n) 
  array.forEach(function (number) {
    if (!(k - number in hash)) {
      hash[ number ] = k - number;
    } else {
      pairs.push([number, k - number]);
    }
  });

  return pairs;
};

File: array-pair-sum/array-pair-sum.rb
def arrayPairSum number, array

  pairs = []
  
  array.each do |a|
      pairs.push([a, a]) if number - a == a && array.count(a) > 1 && !pairs.include?([a, a]) 
      pairs.push([a, number - a]) if number - a != a && array.include?(number - a) && !pairs.include?([number - a, a])
  end
  
  pairs
end
